src/kernel.cr

```diff
@@ -123,24 +123,31 @@ end
 
 # :nodoc:
 module AtExitHandlers
-  @@running = false
-
   def self.add(handler)
```

Comment:
This should probably raise while we're iterating (i.e. `@@handler_iterator != nil`) to stop people modifying the exit handlers in an exit handler.

Reply:
I just tried it in Ruby, it works. We just need to do a loop with an index (in case it doesn't work already).

Reply:
I mean, this works in Ruby:

```ruby
at_exit do
  at_exit do
    puts "BYE"
  end
  puts "EXIT"
end
```

Ouptut:

```
EXIT
BYE
```

Reply:
I don't think it should work. It's confusing and creates for actively confusing code. Let's not follow ruby's example on this.

Reply:
What would be the usecase of having at exit handler adding new at exit handlers?

Reply:
I see nothing wrong with it working. It's intuitive and super easy to implement. But if you think it should be prohibited, it's fine too (but I see nothing wrong with it).

Reply:
For the record, in ruby the nested at_exit handlers are run just after the handler around them:
```ruby
at_exit do
  puts "first"
end

at_exit do
  at_exit do
    puts "> nested"
  end
  puts "second"
end

at_exit do
  puts "third"
end
```
Gives:
```
third
second
> nested
first
```
---
Not sure it can be useful for anything with this behavior...

Reply:
Yes, `at_exit` adds a handler to the end of the list, and to traverse them we should use `pop?`. Super simple. Right now it uses `reverse_each`, but a change in my local copy to use `pop?` made it work.

Reply:
I agree that this is how it should work if we want this to work.

However I don't see the usecase, and so I'd rather be more restrictive until we knew what the usecases were. This is because it creates a far more complex and dynamic control flow through the handlers which may be hard to track and hard to debug. So i'd err on the side of saying "no".

Reply:
Sounds good. We can wait until someone complaints with a real use case.

Reply:
That being said, the `pop?` code is robust whether we disallow this or not, and is smaller and easier to read. If we decide to enable it it should be as easy as removing a `raise ... if` line of code.